controller/integration/test/controlflow/controlflow_test.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	resultObjAccessor, err := meta.Accessor(resultObj)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We verify, that the DeletionTimestamp has not been set.
+	if resultObjAccessor.GetDeletionTimestamp() != nil {
+		t.Fatalf("DeletionTimestamp != nil, want nil")
+	}
+
+	// We define which finalizers we currently expect.
+	expectedFinalizers := []string{
+		testFinalizer,
+	}
+
+	// We verify, that our finalizer is still set.
+	if !reflect.DeepEqual(resultObjAccessor.GetFinalizers(), expectedFinalizers) {
+		t.Fatalf("finalizers == %v, want %v", resultObjAccessor.GetFinalizers(), expectedFinalizers)
+	}
+
+	// We set ThrowError to true, this causes the resource to always throw an error
+	// and should therefor prevent thr removal of our finalizer.
+	tr.ThrowError(true)
+
+	// We delete the object now.
+	err = testWrapper.DeleteObject(objName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We use backoff with the absolute maximum amount:
+	// 10 second ResyncPeriod + 2 second RateWait + 2 second for safety.
+	// The controller should get the deletion event immediatly but not remove the
+	// finalizer because of the error we throw in our resource.
+	//
+	// 		EnsureCreated: 2, EnsureDeleted: 1
+	//
+	// The controller should also reconcile once in this period. (The other
+	// finalizer is still set, so we reconcile.)
+	//
+	// 		EnsureCreated: 2, EnsureDeleted: 2
+	//
+	operation = func() error {
+		if tr.CreateCount() != 2 {
+			return microerror.Maskf(countMismatchError, "EnsureCreated was hit %v times, want %v", tr.CreateCount(), 2)
+		}
+		if tr.DeleteCount() != 2 {
+			return microerror.Maskf(countMismatchError, "EnsureDeleted was hit %v times, want %v", tr.DeleteCount(), 2)
+		}
+		return nil
+	}
+	err = backoff.Retry(operation, newConstantBackoff(uint64(14)))
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We get the object after the controller has handled the deletion event.
+	resultObj, err = testWrapper.GetObject(objName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	resultObjAccessor, err = meta.Accessor(resultObj)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We verify, that our object still exists, but has a DeletionTimestamp set.
+	if resultObjAccessor.GetDeletionTimestamp() == nil {
+		t.Fatalf("DeletionTimestamp == nil, want non-nil")
+	}
+
+	// We define which finalizers we currently expect.
+	expectedFinalizers = []string{
+		testFinalizer,
+	}
+
+	// We verify, that our finalizer is still set.
+	if !reflect.DeepEqual(resultObjAccessor.GetFinalizers(), expectedFinalizers) {
+		t.Fatalf("finalizers == %v, want %v", resultObjAccessor.GetFinalizers(), expectedFinalizers)
+	}
+
+	// We set ThrowError to false, our finalizer should be removed with the next
+	// reconciliation now.
+	tr.ThrowError(false)
```

Comment:
IMO we can drop the argument and only call `ThrowError` when we want the resource to throw an error. So the second time here is not necessary to call it at all.

Reply:
I added the explicit argument for the toggle, so we can add some more complex scenarios when we have a working solution for the cancel context stuff.

I didn't want to add the current solution as it isn't working as you reported in https://github.com/giantswarm/giantswarm/issues/3018